libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_pthread.c

```diff
@@ -364,6 +411,182 @@ TEST( Full_POSIX_PTHREAD, pthread_create_join )
 
 /*-----------------------------------------------------------*/
 
+TEST( Full_POSIX_PTHREAD, pthread_detach_before_running )
+{
+    int iThreadCreateStatus = 0, iStatus = 0;
+    pthread_t xNewThread = NULL;
+    pthread_barrier_t xBarrier = { 0 };
+
+    /* Initialize a pthread barrier for two threads. */
+    iStatus = pthread_barrier_init( &xBarrier, NULL, 2U );
+    TEST_ASSERT_EQUAL_INT( 0, iStatus );
+
+    if( TEST_PROTECT() )
+    {
+        /* Suspend the scheduler to prevent a context switch. */
+        vTaskSuspendAll();
+        {
+            /* Create a new thread within a critical section. By default a joinable thread is created. */
+            iThreadCreateStatus = pthread_create( &xNewThread, NULL, prvBarrierThread, ( void * ) &xBarrier );
+
+            /* Detach the thread before the new thread starts execution. */
+            if( iThreadCreateStatus == 0 )
+            {
+                iStatus = pthread_detach( xNewThread );
+            }
+        }
+        ( void ) xTaskResumeAll();
+
+        TEST_ASSERT_EQUAL_MESSAGE( 0, iThreadCreateStatus, "Thread creation failed." );
+        TEST_ASSERT_EQUAL_MESSAGE( 0, iStatus, "Thread detach failed" );
+
+        /*
+         * Wait on the barrier for new thread to start execution.
+         * Once barrier is released, new thread exits and releases resource by itself.
+         */
+        iStatus = pthread_barrier_wait( &xBarrier );
+
+        if( ( iStatus != 0 ) && ( iStatus != PTHREAD_BARRIER_SERIAL_THREAD ) )
+        {
+            TEST_FAIL_MESSAGE( "Failed to release the barrrier for detached thread." );
+        }
+    }
+
+    /* Destroy the barrier. */
+    ( void ) pthread_barrier_destroy( &xBarrier );
+}
+
+/*-----------------------------------------------------------*/
+
+TEST( Full_POSIX_PTHREAD, pthread_detach_while_running )
+{
+    int iStatus = 0;
+    pthread_t xNewThread = NULL;
+    pthread_barrier_t xBarrier = { 0 };
+
+    /*Initialize a pthread barrier for two threads. */
+    iStatus = pthread_barrier_init( &xBarrier, NULL, 2U );
+    TEST_ASSERT_EQUAL_INT( 0, iStatus );
+
+    if( TEST_PROTECT() )
+    {
+        /* Create a thread with default attributes. By default thread is created as joinable.
+         * Upon creation new thread waits for the current thread to join the barrier.
+         */
+        iStatus = pthread_create( &xNewThread, NULL, prvBarrierThread, ( void * ) &xBarrier );
+        TEST_ASSERT_EQUAL_INT_MESSAGE( 0,
+                                       iStatus,
+                                       "Thread creation failed!" );
+
+        /* Detach the new thread using pthread_detach(). */
+        iStatus = pthread_detach( xNewThread );
```

Comment:
How are we making sure that the thread `prvBarrierThread` is running at this time (as the name of the test suggests)?

Reply:
In fact, the way to test this to call `pthread_detach()` from running thread. Have renamed `pthread_detach_self` test case to the `pthread_detach_while_running`